framework/validators/Validator.php

```diff
@@ -102,6 +102,10 @@ class Validator extends Component
      */
     public $attributes = [];
     /**
+     * @var array cleaned attribute names with "!" characters in the beginning removed
+     */
+    protected $attributeNames = [];
```

Comment:
property should be private as there is a getter and a setter to work with it, there is no need to make it protected.